src/libxtp/threecenter_gwbse.cc

```diff
@@ -149,16 +200,127 @@ void TCMatrix_gwbse::FillBlock(std::vector<Eigen::MatrixXd>& block,
+ */
+void TCMatrix_gwbse::MultiplyRightWithAuxMatrixOpenMP(
+    const Eigen::MatrixXd& matrix) {
+#pragma omp parallel for
+  for (int i_occ = 0; i_occ < _mtotal; i_occ++) {
+    Eigen::MatrixXd temp = _matrix[i_occ] * matrix;
+    _matrix[i_occ] = temp;
+  }
   return;
 }
 
+#if defined(USE_CUDA)
+/*
+ * Convolution of the GW shell with the DFT orbital coefficients using an Nvidia
+ * GPU. The Cuda device behaves like a server that is receiving matrix-matrix
+ * multiplications from a single stream (an Nvidia queue) and handle them
+ * in an asynchronous way. It performs the following operations when recieving a
+ * request:
+ *  1. Check that there is enough space for the arrays
+ *  2. Allocate memory for each matrix
+ *  3. Copy the matrix to the allocated space
+ *  4. Perform the matrix multiplication
+ *  5. Return the result matrix
+ * The Cuda device knows to which memory address it needs to copy back the
+ * result. see: https://docs.nvidia.com/cuda/cublas/index.html#thread-safety2
+ */
+std::vector<Eigen::MatrixXd> TCMatrix_gwbse::FillBlockCUDA(
+    const std::vector<Eigen::MatrixXd>& symmstorage,
+    const std::array<CudaMatrix, 2>& cuda_matrices,
+    std::array<CudaMatrix, 3>& cuda_inter_matrices,
+    const AOShell& shell) const {
+
+  std::vector<Eigen::MatrixXd> block = std::vector<Eigen::MatrixXd>(
+      _mtotal, Eigen::MatrixXd::Zero(_ntotal, shell.getNumFunc()));
+
+  try {
+    CudaPipeline cuda_pip;
+    const CudaMatrix& cuma_A = cuda_matrices[0];
+    const CudaMatrix& cuma_C = cuda_matrices[1];
+    CudaMatrix& cuma_B = cuda_inter_matrices[0];
+    CudaMatrix& cuma_X = cuda_inter_matrices[1];
+    CudaMatrix& cuma_Y = cuda_inter_matrices[2];
+
+    int dim = static_cast<int>(symmstorage.size());
+    for (int k = 0; k < dim; ++k) {
+      const Eigen::MatrixXd& matrix = symmstorage[k];
+      cuma_B.copy_to_gpu(matrix.selfadjointView<Eigen::Lower>());
+      cuda_pip.gemm(cuma_A, cuma_B, cuma_X);
+      cuda_pip.gemm(cuma_X, cuma_C, cuma_Y);
+      Eigen::MatrixXd threec_inMo = cuma_Y;
+      for (int i = 0; i < threec_inMo.cols(); ++i) {
+        block[i].col(k) = threec_inMo.col(i);
+      }
+    }
+  } catch (const std::runtime_error& error) {
+    XTP_LOG_SAVE(logDEBUG, _log)
+        << TimeStamp() << " FillBlockCUDA failed due to: " << error.what()
+        << flush;
+    throw;
+  }
+  return block;
+}
+
+std::array<CudaMatrix, 2> TCMatrix_gwbse::SendDFTMatricesToGPU(
+    const Eigen::MatrixXd& dft_orbitals) const {
+  const Eigen::MatrixXd dftm =
+      dft_orbitals.block(0, _mmin, dft_orbitals.rows(), _mtotal);
+  const Eigen::MatrixXd dftn =
+      dft_orbitals.block(0, _nmin, dft_orbitals.rows(), _ntotal);
+
+  // Smart Pointers to the cuda arrays
+  CudaPipeline cuda_pip;
+  const cudaStream_t& stream = cuda_pip.get_stream();
+
+  return {CudaMatrix{dftn.transpose(), stream}, CudaMatrix{dftm, stream}};
+}
+
+std::array<CudaMatrix, 3> TCMatrix_gwbse::CreateIntermediateCudaMatrices(
```

Comment:
replace the `const Eigen::MatrixXd& dft_orbitals` with `int basissize` because that is all you need

Reply:
done